File: python/maze.py
from collections import defaultdict
from readFromGraph import *

class Node:
    def __init__(self, n=None, s=None, w=None, e=None, u=None, d=None) -> None:
        self.north = n
        self.south = s
        self.west = w
        self.east = e
        self.up = u
        self.down = d

'''
class Graph:
    def __init__(self, n) -> None:
        self.graph = {i + 1: [] for i in range(n)}
    
    def addNode(self, level, n, s, w, e, u, d):
        node = Node(n, s, w, e, u, d)
        self.graph[level].append(node)
    
    def printLevel(self, level):
        return self.graph[level]
'''   

def direction(bits):
    directions = {0: 'N', 1:'E', 2:'S', 3:'W', 4:'U', 5:'D'}
    res = ''
    for num, b in enumerate(bits):
        if int(b):
            res += directions[num]
    return res

'''
def bfs(graph, vertex):
    visited = set()
    queue = [vertex]
    visited.add(vertex)
    while queue:
        v = queue.pop(0)
        print(v)
        for edges in graph[v]:
            if edges not in visited:
                visited.add(edges)
                queue.append(edges)
'''

def helper(v, vSet, p, dir, q):
    if v not in vSet:
        vSet.add(v)
        p.append(dir)
        q.append((v, p))

def bfs(start, end):
    visited = set()
    queue = [(start, [])]
    visited.add(start)

    while queue:
        node, path = queue.pop(0)
        # print(path)
        if node == end:
            return path
        visit = None
        if node.north:
            # print('north')
            visit = node.north
            helper(visit, visited, path.copy(), 'N', queue)
        if node.south:
            # print('south')
            visit = node.south
            helper(visit, visited, path.copy(), 'S', queue)
        if node.west:
            # print('west')
            visit = node.west
            helper(visit, visited, path.copy(), 'W', queue)
        if node.east:
            # print('east')
            visit = node.east
            helper(visit, visited, path.copy(), 'E', queue)
        if node.up:
            # print('up')
            visit = node.up
            helper(visit, visited, path.copy(), 'U', queue)
        if node.down:
            # print('down')
            visit = node.down
            helper(visit, visited, path.copy(), 'D', queue)

if __name__ == "__main__":
    binaryGraph, levels = readGraph('/Users/emanuelaseghehey/Development/Itsy-Bitsy-Spider-algo/textfiles/tiny-maze.txt')
    # print(binaryGraph)
    graph = {i: [] for i in range(levels)}
    lvl = 0

    for k, v in binaryGraph.items():
        for i in range(len(v)):
            temp = []
            for j in range(len(v[i])):
                temp.append(Node())
            graph[lvl].append(temp)
        lvl += 1

    # print(direction(4, binaryGraph[1][0][0]))
    dir = {i: [] for i in range(levels)}
    lvl = 0
    for k, v in binaryGraph.items():
        for i in range(len(v)):
            temp = []
            for j in range(len(v[i])):
                temp.append(direction(v[i][j]))
            dir[lvl].append(temp)
        lvl += 1

    # print(dir)
    # print(binaryGraph)
    for lvl, vals in dir.items():
        for i in range(len(vals)):
            for j in range(len(vals[i])):
                for w in vals[i][j]:
                    if w == 'N':
                        graph[lvl][i][j].north = graph[lvl][i-1][j]
                    elif w == 'S':
                        graph[lvl][i][j].south = graph[lvl][i + 1][j]
                    elif w == 'W':
                        graph[lvl][i][j].west = graph[lvl][i][j-1]
                    elif w == 'E':
                        graph[lvl][i][j].east = graph[lvl][i][j+1]
                    elif w == 'U':
                        graph[lvl][i][j].up = graph[lvl + 1][i][j]
                    else:
                        graph[lvl][i][j].down = graph[lvl - 1][i][j]

    # print(graph)
    f_analysis = open('analysis', 'w')
    for k, v in dir.items():
        for vals in v:
            f_analysis.write('    '.join(vals) + '\n')
        f_analysis.write('\n')
    f_analysis.close()
    # print(dir[3][2][2])
    # print(graph.keys())
    f_out = open('output.txt', 'w')
    f_out.write(' '.join(bfs(graph[0][0][0], graph[2][2][2])))
    f_out.close()
    # print(bfs(graph[1][0][0], graph[1][1][0]))

    # print(graph[0][0][0])

File: cpp final/maze.cpp
#include <fstream>
#include <string>
#include <utility>
#include <stdexcept>
#include <vector>
#include <sstream>
#include <iostream>
#include <map>
#include <tuple>
#include <queue>

int totalsize = 0;

struct Node {
    Node* North;
    Node* South;
    Node* East;
    Node* West;
    Node* Up;
    Node* Down;
    int id;

    Node() {
        North = nullptr;
        South = nullptr;
        East = nullptr;
        West = nullptr;
        Up = nullptr;
        Down = nullptr;
    }
};

std::string getDirections(const std::string& binaryString){    
    std::string result;
    for (int i = 0; i < binaryString.length(); i++){
        if (binaryString[i] == '1'){
            switch(i){
                case 0:
                    result += 'N';
                    break;
                case 1:
                    result += 'E';
                    break;
                case 2:
                    result += 'S';
                    break;
                case 3:
                    result += 'W';
                    break;
                case 4:
                    result += 'U';
                    break;
                case 5:
                    result += 'D';
                    break;
            }
        }
    }
    return result;
}

// auto readInput(const std::string& filename){
//     std::ifstream file(filename);
//     std::string results;

//     if (!file.is_open()){
//         throw std::runtime_error("cannot read file");
//     }

//     std::tuple <int, int, int> size;
//     std::tuple <int, int, int> start;
//     std::tuple <int, int, int> end;

// }

void addToTuple(std::ifstream& file, std::tuple<int, int, int>& tup){
    file >> std::get<0>(tup);
    file >> std::get<1>(tup);
    file >> std::get<2>(tup);
}

void helper(Node v, std::vector<bool>& visited, std::vector<std::string> path, std::string direction, std::queue<std::pair<Node, std::vector<std::string>>>& q){

    if (visited[v.id] == false) {
        visited[v.id] = true;
        path.push_back(direction);
        std::pair<Node, std::vector<std::string>> addToQueue(v, path);
        q.push(addToQueue);
    }
}

std::string bfs(Node start, Node end){
    // create visit
    std::vector<bool> visited;
    visited.resize(totalsize, false);
    std::queue<std::tuple<Node, std::vector<std::string>>> queue;
    visited[start.id] = true;

    while(!queue.empty()){
        std::tuple<Node, std::vector<std::string>> nodeTuple = queue.front();
        queue.pop();
    }
}

int main(){
    
    std::ifstream file("/Users/emanuelaseghehey/Development/Itsy-Bitsy-Spider-algo/textfiles/tiny-maze.txt");
    std::string results;

    if (!file.is_open()){
        std::cout << "cannot";
        throw std::runtime_error("cannot read file");
    }

    std::tuple <int, int, int> size; // l, r, c
    std::tuple <int, int, int> start;
    std::tuple <int, int, int> end;


    std::vector<std::vector<std::vector<Node>>> graph;    
    std::vector<std::vector<std::vector<std::string>>> binaryGraph;

    addToTuple(file, size);
    addToTuple(file, start);
    addToTuple(file, end);

    // creating both graphs, actual and helper
    for (int i = 0; i < (std::get<0>(size)*std::get<1>(size)); i=i + std::get<1>(size)){
        std::vector<std::vector<Node>> buildNodeGraph;
        std::vector<std::vector<std::string>> current_level;
        for (int j = 0; j < std::get<1>(size); j++){
            std::vector<Node> buildNodes;
            std::vector<std::string> current_row;
            for (int k = 0; k < std::get<2>(size); k++){
                std::string temp;
                file >> temp;
                current_row.push_back(getDirections(temp));
                Node cur;
                cur.id = totalsize;
                totalsize += 1;
                buildNodes.push_back(cur);
            }
            buildNodeGraph.push_back(buildNodes);
            current_level.push_back(current_row);
        }
        graph.push_back(buildNodeGraph);
        binaryGraph.push_back(current_level);
    }
    
    for (int i = 0; i < binaryGraph.size(); i++){
        for (int j = 0; j < binaryGraph[i].size(); j++){
            for (int k = 0; k < binaryGraph[i][j].size(); k++){
                std::string current_string = binaryGraph[i][j][k];
                for (int w = 0; w < current_string.length(); w++){
                    switch(current_string[w]){
                        case 'N':
                            graph[i][j][k].North = &graph[i][j-1][k];
                            break;
                        case 'S':
                            graph[i][j][k].South = &graph[i][j+1][k];
                            break;
                        case 'W':
                            graph[i][j][k].West = &graph[i][j][k-1];
                            break;
                        case 'E':
                            graph[i][j][k].East = &graph[i][j][k+1];
                            break;
                        case 'U':
                            graph[i][j][k].Up = &graph[i+1][j][k];
                            break;
                        case 'D':
                            graph[i][j][k].Down = &graph[i-1][j][k];
                            break;
                    }
                }
            }
        }
    }

    // test
    // for (int i = 0; i < binaryGraph.size(); i++){
    //     for (int j = 0; j < binaryGraph[i].size(); j++){
    //         for (int k = 0; k < binaryGraph[i][j].size(); k++){
    //             std::cout << graph[i][j][k].id << ' ';
    //         }
    //         std::cout << std::endl;
    //     }
    //     std::cout << std::endl;
    // }

    // std::cout << graph[0][0][0].Up << std::endl;
    // std::cout << &graph[1][0][0] << std::endl;

    std::vector<bool> visited;
    visited.resize(totalsize, false);
    std::queue<std::pair<Node, std::vector<std::string>>> queue;
    std::vector<std::string> start_path;

    std::pair<Node, std::vector<std::string>> p(graph[0][0][0], start_path);
    queue.push(p);
    visited[graph[0][0][0].id] = true;

    while(!queue.empty()){

        std::pair<Node, std::vector<std::string>> current = queue.front();
        queue.pop();

        Node cur = current.first;
        std::vector<std::string> path = current.second;

        // test
        // for (int i = 0; i < path.size(); i ++){
        //         std::cout << path[i] << ' ';
        // }
        // std::cout << std::endl;

        if (cur.id == graph[2][2][2].id){
            for (int i = 0; i < path.size(); i ++){
                std::cout << path[i] << ' ';
            }
            // break;
        }

        if (cur.North != nullptr){
            helper(*cur.North, visited, path, "N", queue);
        }

        if (cur.South != nullptr){
            // come back
            helper(*cur.South, visited, path, "S", queue);
        }

        if (cur.East != nullptr){
            helper(*cur.East, visited, path, "E", queue);
        }

        if (cur.West != nullptr){
            helper(*cur.West, visited, path, "W", queue);
        }

        if (cur.Up != nullptr){
            helper(*cur.Up, visited, path, "U", queue);
        }

        if (cur.Down != nullptr){
            helper(*cur.Down, visited, path, "D", queue);
        }
    }

    return 0;
}
